### DISCUSSION TITLE
Heterogenous GNN

### DISCUSSION THREAD

**Discussion by AlizadehF** [pyg-team/pytorch_geometric#9250] [Q&A]
Hello everybody,

How HeteroGNN learn different types of edge with a different feature length? 
`conv = HeteroConv({
                edge_type: SAGEConv((-1, -1), hidden_channels, aggr=aggr)
                for edge_type in metadata[1]})
            self.convs.append(conv)`

after `convs.append` , is the heteroconv convert to homogenous graph??
if the features length are 3,5,10, after conv.append and aggregations of them the edge/node embedding length is 10? it means the bigest length is the length of embedding? it produce large of zero value?!

**Comment by rusty1s** [comment]
Sorry, I don't fully understand yet. `HeteroConv` expects a dictionary of node features/edge indices for each node/edge type, and computes new node features and returns them as a dictionary again. There is no hetero<>homogeneous conversion in this case. However, after the GNN is applied, all node features will have the same embedding dimension (i.e., `hidden_channels`), while this is not a requirement for input features.

**Comment by AlizadehF** [comment]
![Untitled](https://github.com/pyg-team/pytorch_geometric/assets/73420058/dfa18ffb-9635-42e9-b5e9-f96c3e1634dc)

Thanks,
I mean, how long is the embedding after merging features of different lengths
As drawn in the figure. Finally, how is node embedding done?

**Reply by rusty1s** [reply]
Neighbors of different node types get transformed into equally sized features (e.g., a parameter matrix for each node type), so that nodes across different types can be successfully aggregated. The size of the final features is defined via the `out_channels` of the GNN layers.

**Comment by AlizadehF** [comment]
I have different type of edges. nodes are the same. May I define the node classification task with different edge type? I mean, I aggregate the different types of edge feature length and add with target node's feature?

**Reply by rusty1s** [reply]
Yes, that is possible, see `RGCN` for example.
